src/main/scala/uk/gov/nationalarchives/tdr/api/service/ConsignmentStatusService.scala

```diff
@@ -78,6 +81,27 @@ class ConsignmentStatusService(
     )
   }
 
+  def updateMetadataConsignmentStatus(consignmentId: UUID, statusTypes: List[String]): Future[List[Int]] = {
+    fileStatusRepository
+      .getFileStatus(consignmentId, statusTypes.toSet)
+      .flatMap(fileStatuses => {
+        statusTypes
+          .map(statusType => {
+            val fileStatusesForType = fileStatuses.filter(_.statustype == statusType)
+            val statusValue = if (fileStatusesForType.count(_.value == NotEntered) == fileStatusesForType.length) {
+              NotEntered
+            } else if (fileStatusesForType.exists(_.value == Incomplete)) {
+              Incomplete
+            } else {
+              Completed
+            }
+            val input = ConsignmentStatusInput(consignmentId, statusType, Option(statusValue))
+            updateConsignmentStatus(input)
+          })
+          .sequence
+      })
+  }
+
```

Comment:
I just realized, if we get the correct consignment status from the SQL query itself, it might be a bit faster.
What do you think?

I tried to create one but I haven't tested it with a different set of data:

```

SELECT "StatusType", (CASE WHEN COUNT(CASE WHEN "Value" = 'Incomplete' THEN 1 ELSE NULL END) > 0 THEN 'Incomplete'
						   ELSE (CASE WHEN COUNT(CASE WHEN "Value" = 'Completed' THEN 1 ELSE NULL END) > 0 THEN 'Completed' ELSE 'NotEntered' END)
						   END) as Value						   						   
FROM "FileStatus" fs JOIN "File" f ON f."FileId" = fs."FileId"
WHERE f."ConsignmentId" = 'b7cb69b7-3df3-4e00-886b-1261b942d6ff'
GROUP BY "StatusType" HAVING "StatusType" IN ('ClosureMetadata', 'DescriptiveMetadata');
```

Reply:
I think it might be best to leave this for now. We've had raw SQL before and it has caused problems and I'm not sure if this will be too much faster. If there are problems that come out of the load testing then we can re-visit this.